Improve Icon tests and documentation (#571)
jpmorganchase/uitk: packages/icons/stories/icon.doc.stories.mdx

<Meta
  title="Documentation/Icons/Icon"
  component={Icon}
  parameters={{
    viewMode: "docs",
  }}
/>

# Icon

Icon is an important visual representation of an application, a capability, or some other concept or specific entity with meaning on a interface. Icons help users to recognize and understand a concept or an action regardless of linguistic or cultural boundaries.

If an icon is interactive, use [Button](?path=/docs/documentation-core-button--button-variants) (`CTA`, `Primary` or `Secondary` variant).

<Canvas>
  <Story id="icons-icon--toolkit-icon" />
</Canvas>

## Use Icon when...

- You need a visual representation of a concept.
- An icon doesn't need to be interactive.

## Features

We have a default icon set that starts from 12px for high- and medium density (MD), then 14px for low density (LD), and 16px for touch density (TD). (Since our base size of high density (HD)/default is 10px but we don’t allow icons size smaller than 12px, hence why here the HD/default will start from 12px, and its 2X is 20px, 3X is 30px etc.)
You have more size choices by adding multiplies to 2x, 3x, 4x etc.

<Canvas>
  <Story id="icons-icon--toolkit-icon-multiple-sizes" />
</Canvas>

> We provide two types of icons: default and solid. Some special icons may be excluded from it like `plus` and `loader`.

Reviewer: At the moment the second sentence here is a little unclear. We should provide a full list of icons that are excluded  - 'Some' is too vague - and explain why they are excluded/special.

Reviewer 2: @pseys Maybe you can provide more context here

Author: I checked with @wennazhang22 for context and updated this paragraph

Reviewer 3: Has this question been resolved now? Sorry for the delay, I must've missed a notification. Its not that they're special, they just don't have solid versions because it's not relevant/required. For example, the plus icon doesn't have a 'fillable' aspect so can't have a solid version.

Reviewer 2: @sion-docs Does the content still need to be changed?

Reviewer: See the most recent comment below, Fernanda has changed it.

<Canvas>
  <Story id="icons-icon--toolkit-types" />
</Canvas>

> To override the default color of the icon, you can use the `color` css variable.

Reviewer 2: I'm thinking we probably want to refer to the whole CSS variable. Thoughts?

<Canvas>
  <AddDocumentIcon
    style={{ "--uitkIcon-color": "var(--uitk-status-info-foreground)" }}
  />
</Canvas>

> You can supply custom icons as `children` of the `Icon` component by using an `<svg>`. Whenever you add a custom icon you need to pass an `aria-label` to the Icon.

Reviewer 2: I still feel like this needs clarification. What does using an <SVG> mean?

Should this be something more like:

```suggestion
You can supply custom icons as `children` of the `Icon` component. `Icon` has the same API as the native `<svg>` element. Whenever you add a custom icon you need to pass an `aria-label` to the Icon.
```

Author: I'll update this, @sion-docs / @joshwooding would you say the next paragraph makes sense here or should i move it to accessibility? (or some content in both)

Reviewer 2: I think it's fine for it to be here. Whether that is achieved by putting it directly here or linking to it and moving it to the accessibility section I don't have a preference

Reviewer: I'm happy to go with Josh's judgment.

Author: happy to leave it here :)

> 

Reviewer: Changed to make the tone more active/2nd person. Should aria-label be `monospace`?

> You can supply custom icons as `children` of the `Icon` component by using an `<svg>`. It is necessary to pass an aria-label to the Icon whenever adding a custom icon. 

Author: @sion-docs fyi. I just added another bit about labels here, if you could check if it makes sense that would be great :)

Reviewer: I have updated the text and made another commit, see below.

<Canvas>
  <Story id="icons-icon--custom-svg-icon" />
</Canvas>

### Icons List

<Canvas>
  <Story id="icons-icon--all-icons-with-search" />
</Canvas>

## Configuring Icon

### API

```
import { Icon } from "@jpmorganchase/uitk-icons";
```

### Props

<ArgsTable of={Icon} />

### Anatomy

<Canvas>
  <ComponentAnatomy>
    <AddDocumentIcon />
  </ComponentAnatomy>
</Canvas>

### Classes

<CSSClassTable of={Icon} />

## Characteristics Used

<CharacteristicUsage of={Icon} />

## Accessibility

> To ensure that icon-only buttons are ADA compliant, when icon is inside a button you must display the button’s label in a supporting Tooltip that is revealed on mouse hover or keyboard focus.

Reviewer 2: I would argue this is not specific to Icon and shouldn't be here. Seems like something that should be on the Button page and potentially linked to

import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs";
import { Icon, AddDocumentIcon } from "@jpmorganchase/uitk-icons";
import { withFlexGap } from "docs/decorators/withFlexGap";
import { ComponentAnatomy } from "docs/components/ComponentAnatomy";

import {
  CharacteristicUsage,
  CSSClassTable,
} from "css-variable-docgen-components";

<Meta
  title="Documentation/Icons/Icon"
  component={Icon}
  parameters={{
    viewMode: "docs",
  }}
/>

# Icon

Icon is an important visual representation of an application, a capability, or some other concept or specific entity with meaning on a interface. It is expected to help user to recognize and understand a content or an action regardless languages or culture boundaries.

> We have 3 sizes of icon set for users to choose in terms of different platforms or device size.

Reviewer 4: I guess we'll need to update the size part of this

import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs";
import { Icon, AddDocumentIcon } from "@jpmorganchase/uitk-icons";
import { withFlexGap } from "docs/decorators/withFlexGap";
import { ComponentAnatomy } from "docs/components/ComponentAnatomy";

import {
  CharacteristicUsage,
  CSSClassTable,
} from "css-variable-docgen-components";

<Meta
  title="Documentation/Icons/Icon"
  component={Icon}
  parameters={{
    viewMode: "docs",
  }}
/>

# Icon

Icon is an important visual representation of an application, a capability, or some other concept or specific entity with meaning on a interface. It is expected to help user to recognize and understand a content or an action regardless languages or culture boundaries.

If an icon is interactive, use [Button](?path=/docs/documentation-core-button--button-variants) (`CTA`, `Primary` or `Secondary` variant).

<Canvas>
  <Story id="icons-icon--toolkit-icon" />
</Canvas>

## Use Icon when...

- You need a visual representation of a concept.
- An icon doesn't need to be interactive.

## Features

> We have default icon set that starts from 12px for HD & MD, then 14px for LD, 16px for TD. (Since our base size of HD/default is 10px, but we don’t allowing icon size that is smaller then 12px, hence why here, the HD/default will start from 12px, and its 2X is 20px, 3X is 30px etc.)

Reviewer 4: "We have **a** default icon set"

Have more size choices by adding multiplies to 2x, 3x, 4x etc.

<!--  <Canvas><Story id="" /></Canvas> -->

> We provide 2 types of icons: default and solid. Some special icon may exclueded from it like plus and loader

Reviewer 4: excluded typo

<Canvas>
  <Story id="icons-icon--toolkit-types" />
</Canvas>

### Icons List

<Canvas>
  <Story id="icons-icon--all-icons-with-search" />
</Canvas>

## Configuring Icon

### API

```
import { Icon } from "@jpmorganchase/uitk-icons";
```

### Props

<ArgsTable of={Icon} />

## Anatomy

We have default and solid icons.  

> You can supply custom icons as `children` of the `Icon` component, by using `<svg>`.

Reviewer 2: This doesn't fall under Anatomy to me.
Also I believe the custom icon guidance needs to change after the other PR.

<Meta
  title="Documentation/Icons/Icon"
  component={Icon}
  parameters={{
    viewMode: "docs",
  }}
/>

# Icon

Icon is an important visual representation of an application, a capability, or some other concept or specific entity with meaning on a interface. Icons help users to recognize and understand a concept or an action regardless of linguistic or cultural boundaries.

If an icon is interactive, use [Button](?path=/docs/documentation-core-button--button-variants) (`CTA`, `Primary` or `Secondary` variant).

<Canvas>
  <Story id="icons-icon--toolkit-icon" />
</Canvas>

## Use Icon when...

- You need a visual representation of a concept.
- An icon doesn't need to be interactive.

## Features

We have a default icon set that starts from 12px for high and medium density (MD), then 14px for low density (LD), and 16px for touch density (TD). (Since our base size of high density (HD)/default is 10px but we don’t allow icons size smaller than 12px, hence why here the HD/default will start from 12px, and its 2X is 20px, 3X is 30px etc.)
You have more size choices by adding multiplies to 2x, 3x, 4x etc.

<Canvas>
  <Story id="icons-icon--toolkit-icon-multiple-sizes" />
</Canvas>

> We provide two types of icons: default and solid. Most of our icons will have the two versions, except cases in which the icon shape is too narrow to have a solid version like `plus` and `loader`.

Author: @pseys

Reviewer 3: Great, thank you.

Reviewer: This makes sense.

<Canvas>
  <Story id="icons-icon--toolkit-types" />
</Canvas>

To override the default color of the icon, you can use the `--uitkIcon-color` CSS variable.

<Canvas>
  <AddDocumentIcon
    style={{ "--uitkIcon-color": "var(--uitk-status-info-foreground)" }}
  />
</Canvas>

> You can supply custom icons as `children` of the `Icon` component. `Icon` has the same API as the native `<svg>` element. Whenever you add a custom icon you need to pass an `aria-label` to the Icon.

Reviewer 2: ```suggestion
If you need a custom SVG icon you can use the `Icon` component. `Icon` has the same API as the native `<svg>` element. Whenever you add a custom icon you need to pass an `aria-label` to the Icon.
```

Reviewer 2: cc @sion-docs

Reviewer: Reads fine to me

## Features

> We have a default icon set that starts from 12px for high- and medium density (MD), then 14px for low density (LD), and 16px for touch density (TD). (Since our base size of high density (HD)/default is 10px but we don’t allow icons size smaller than 12px, hence why here the HD/default will start from 12px, and its 2X is 20px, 3X is 30px etc.)

Reviewer 2: ```suggestion
We have a default icon set that starts from 12px for high and medium density (MD), then 14px for low density (LD), and 16px for touch density (TD). (Since our base size of high density (HD)/default is 10px but we don’t allow icons size smaller than 12px, hence why here the HD/default will start from 12px, and its 2X is 20px, 3X is 30px etc.)
```

  />
</Canvas>

> You can supply custom icons as 'children' of the `Icon` component by using an `<svg>`.

Reviewer 2: Just curious about this change. `children` is a special value when dealing with components.

Also @Fercas123 seems like this still needs to be updated

Reviewer: my mistake
